src/tags/object/RichText/model.js

```diff
@@ -0,0 +1,247 @@
+  name: types.identifier,
+  value: types.maybeNull(types.string),
+
+  /** Defines the type of data to be shown */
+  valuetype: types.optional(types.enumeration(["text", "url"]), () => (window.LS_SECURE_MODE ? "url" : "text")),
+
+  /** Wether or not to save selected text to the serialized data */
+  savetextresult: types.optional(types.enumeration(["none", "no", "yes"]), () =>
+    window.LS_SECURE_MODE ? "no" : "yes",
+  ),
+
+  selectionenabled: types.optional(types.boolean, true),
+
+  clickablelinks: false,
+
+  highlightcolor: types.maybeNull(customTypes.color),
+
+  showlabels: types.optional(types.boolean, true),
+
+  encoding: types.optional(types.enumeration(["none", "base64", "base64unicode"]), "none"),
+
+  granularity: types.optional(types.enumeration(["symbol", "word", "sentence", "paragraph"]), "symbol"),
+});
+
+const Model = types
+  .model("RichTextModel", {
+    type: "richtext",
+    _value: types.optional(types.string, ""),
+    _update: types.optional(types.number, 1),
+  })
+  .views(self => ({
+    get hasStates() {
+      const states = self.states();
+      return states && states.length > 0;
+    },
+
+    get completion() {
+      return getRoot(self).completionStore.selected;
+    },
+
+    get regs() {
+      return self.completion.regionStore.regions.filter(r => r.object === self);
+    },
+
+    states() {
+      return self.completion.toNames.get(self.name);
+    },
+
+    activeStates() {
+      const states = self.states();
+      return states ? states.filter(s => s.isSelected && SUPPORTED_STATES.includes(getType(s).name)) : null;
+    },
+  }))
+  .volatile(self => ({
+    rootNodeRef: React.createRef(),
+  }))
+  .actions(self => ({
+    async updateValue(store) {
+      const value = parseValue(self.value, store.task.dataObj);
+
+      if (self.valuetype === "url") {
+        const url = value;
+
+        if (!/^https?:\/\//.test(url)) {
+          const message = [WARNING_MESSAGES.dataTypeMistmatch(), WARNING_MESSAGES.badURL(url)];
+          if (window.LS_SECURE_MODE) message.unshift(WARNING_MESSAGES.secureMode());
+
+          Infomodal.error(message.map(t => <p>{t}</p>));
+          self.setRemoteValue("");
+          return;
+        }
+
+        const response = await fetch(url);
+        const { ok, status, statusText } = response;
+
+        try {
+          if (!ok) throw new Error(`${status} ${statusText}`);
+
+          self.setRemoteValue(await response.text());
+        } catch (error) {
+          Infomodal.error(WARNING_MESSAGES.loadingError(url, error));
+          self.setRemoteValue("");
+        }
+      } else {
+        self.setRemoteValue(value);
+      }
+    },
+
+    setRemoteValue(val) {
+      self.loaded = true;
+
+      if (self.encoding === "base64") val = atob(val);
+      if (self.encoding === "base64unicode") val = Utils.Checkers.atobUnicode(val);
+
+      self._value = val;
+
+      self._regionsCache.forEach(({ region, completion }) => {
```

Comment:
I hope that this cache is redundant in the new engine.
it was used to store regions before the full load by url and create correct spans only after that.